### PR TITLE
Fix Potential Memory Bloats.

### PR DESCRIPTION
**Pull request by 4n4nd**
Use size limited LRUCache dictionaries to keep memory be used constant

### CODE REVIEW

**Comment by 4n4nd** [comment]
Fix for Issue #84

File: flatliners/alertcomparisonscore.py [hunk 1]
```diff
@@ -6,13 +6,13 @@
...
 class AlertComparisonScore(BaseFlatliner):
-    def __init__(self):
+    def __init__(self, max_cache_size: int = 500):
         super().__init__()
 
-        self.score = dict()
-        self.clusters = dict()
-        self.versions = dict()
-        self.alert_deltas = dict()
+        self.score = super().create_cache_dict(maxsize=max_cache_size)
```
Comment (durandom) [inline flatliners/alertcomparisonscore.py]: do we really need `super()` here? I think it's special for the constructor `__init__` call, but other than that it should work with `self.create...`

**Comment by durandom** [comment]
what happens if `max_size` is reached?

File: flatliners/alertfrequencyversion.py [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
-
 class AlertFrequencyVersion(BaseFlatliner):
 
     """
     This class contains the functionalities to calculate
     alert correlation shift over multiple gitversion
     """
 
-    def __init__(self):
+    def __init__(self, max_cache_size: int = 500):
```
Comment (durandom) [inline flatliners/alertfrequencyversion.py]: let's not init these at the constructor. the default can be in the super class

**Review by durandom — CHANGES_REQUESTED** [review]
* default value in base_flatliner is enough
* remove `super()` from init calls

File: flatliners/alertcomparisonscore.py (hunk 1 above)
Comment (4n4nd) [reply flatliners/alertcomparisonscore.py]: Does it make a difference if the method is a static method?

Comment (durandom) [reply flatliners/alertcomparisonscore.py]: dont know. 
I would make it a member method and not a static one - but up to you

**Comment by durandom** [comment]
I think we also rather need a TTLCache with a reasonably high maxsize. Most of the time the maxsize will be determined by the number of clusters.

**Comment by 4n4nd** [comment]
Implemented the TTLcache dicts for `VersionedMetrics`, if this looks good, I will apply the same changes to other flatliners as well

File: flatliners/versionedmetrics.py [hunk 3]
```diff
@@ -1,12 +1,11 @@
 from .baseflatliner import BaseFlatliner
 
-
 class VersionedMetrics(BaseFlatliner):
     def __init__(self):
         super().__init__()
 
         # dict for versions that we already know
-        self.cluster_versions = dict()
+        self.cluster_versions = self.create_cache_dict()
```
Comment (durandom) [inline flatliners/versionedmetrics.py]: please add a comment about the expected size of this dict and how to find out what size is best.
and what is stored in the dict. maybe an example

File: flatliners/baseflatliner.py [hunk 4]
```diff
@@ -41,6 +42,12 @@ def cluster_id(self, x):
     def cluster_version(self, x):
         return self.metric_label(x, 'version')
 
-
-
-
+    @staticmethod
+    def create_cache_dict(maxsize=500, ttl=600):
+        # Here max size is determined by the number of clusters whose
```
Comment (durandom) [inline flatliners/baseflatliner.py]: that's not true. this just creates a cached dict. depends on the caller what he wants to store in this dict

**Review by durandom — CHANGES_REQUESTED** [review]
better comments please

File: flatliners/baseflatliner.py (hunk 4 above)
Comment (4n4nd) [reply flatliners/baseflatliner.py]: `max size is determined` should be *Default value of maxsize was determined*

Comment (durandom) [reply flatliners/baseflatliner.py]: I mean this method doesnt know anything about the number of clusters... so dont put a reference in here.

File: flatliners/versionedmetrics.py (hunk 3 above)
Comment (durandom) [reply flatliners/versionedmetrics.py]: @4n4nd 

> please add a comment about the expected size of this dict and how to find out what size is best.

Comment (durandom) [reply flatliners/versionedmetrics.py]:
```python
 # caches the version for a specific cluster_id, e.g. cluster_versions['_id_123'] = 'version4'
 # the expected size the number of clusters reporting into telemeter
 # promql: count (count by (_id) (cluster_version))
 self.cluster_versions = self.create_cache_dict(max_size = 1000)
```

File: flatliners/baseflatliner.py [hunk 5]
```diff
@@ -41,6 +42,12 @@ def cluster_id(self, x):
     def cluster_version(self, x):
         return self.metric_label(x, 'version')
 
-
-
-
+    @staticmethod
+    def create_cache_dict(maxsize=500, ttl=600):
```
Comment (durandom) [inline flatliners/baseflatliner.py]: let's bump `maxsize` to 2000. That's not a lot from a memory perspective...

**Comment by 4n4nd** [comment]
recheck

**Comment by durandom** [comment]
@goern @fridex you recommend following the `setup.py` pattern for running the tests here too?

https://managesf.thoth-station.ninja/logs/87/87/8167a9f3656e12347baaa051ce8cbb939c684a9b/check/thoth-pytest/9ddbd89/ara-report/result/55980b64-9bc8-4381-88d3-eed46aecd2d8/

**Review by durandom — APPROVED** [review]
great!

now on to the other flatliners :)

**Pull request merged**